### PR TITLE
HPCC-29517 Skip stderr logging if stderr > /dev/null

### PR DESCRIPTION
**Pull request by mckellyln**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

**Comment by mckellyln** [comment]
For discussion ...

File: system/jlib/jlog.cpp [hunk 1]
```diff
@@ -1678,7 +1679,37 @@ void CLogMsgManager::resetMonitors()
...
     removeAllMonitors();
     Owned<ILogMsgFilter> defaultFilter = getDefaultLogMsgFilter();
-    addMonitor(theStderrHandler, defaultFilter);
+
+    // dont add log handler if stderr is redirected to /dev/null ...
+
+    bool okToAddMonitor = true;
+#ifndef _WIN32
+    int fd2 = fileno(stderr);
+    if (fd2 >= 0)
```
Comment (ghalliday) [inline system/jlib/jlog.cpp:1688]: I would extract this into a separate function isRedirectedToDevNull() or something similar.  (Including the windows #ifdef).

**Review by ghalliday — COMMENTED** [review]
@mckellyln looks like a worthwhile change.  One comment about the structure.

File: system/jlib/jlog.cpp (hunk 1 above)
Comment (mckellyln) [reply system/jlib/jlog.cpp]: ok, sure.  You mean like this ?  (static in same file) or something external in jfile etc. that takes a FILE * arg ?
When it looks ok I will squash and create a JIRA etc.

File: system/jlib/jlog.cpp [hunk 2]
```diff
@@ -28,6 +28,7 @@
 #include "jmisc.hpp"
 #include "jprop.hpp"
 #include "lnuid.h"
+#include <sys/stat.h>
```
Comment (ghalliday) [inline system/jlib/jlog.cpp:31]: Probably needs to be in a #ifndef _WIN32

File: system/jlib/jlog.cpp [hunk 3]
```diff
@@ -60,6 +61,38 @@ static FILE *getNullHandle()
 #endif
 }
 
+static bool StdErrIsDevNull()
```
Comment (ghalliday) [inline system/jlib/jlog.cpp:64]: picky: normally have lower case leading character for function names

**Review by ghalliday — COMMENTED** [review]
Looks good.  A couple of comments.  Also needs to be a non-draft PR

**Review by ghalliday — COMMENTED** [review]
Looks good.  A couple of comments.  Also needs to be a non-draft PR (and associated with a jira)

File: system/jlib/jlog.cpp (hunk 2 above)
Comment (ghalliday) [reply system/jlib/jlog.cpp:31]: I take it back - it looks like it is supported by windows.

**Comment by mckellyln** [comment]
Changed from draft to regular PR and added JIRA and attached PR
https://track.hpccsystems.com/browse/HPCC-29517

**Review by ghalliday — APPROVED** [review]
@mckellyln approved, but please rebase onto 9.0.x - I think that is a more appropriate target for an optimization

**Comment by mckellyln** [comment]
Rebased --onto candidate-9.0.x

**Review by jakesmith — APPROVED** [review]
Looks good

**Pull request merged**